### PR TITLE
Fix invalid handle bug happening when `TypeBuilder` type used in exception catch clause

### LINKED ISSUE #106485: PersistedAssemblyBuilder ILGenerator cannot handle catch block with newly created exception type
`ILGenerator.BeginCatchBlock(Type? exceptionType)` doesn't work for a newly generated exception type because the `TypeBuilder` token is not valid and it will not be populated until the assembly saved into a file or stream 

Sampe repro:
```cs
PersistedAssemblyBuilder ab = new PersistedAssemblyBuilder(new AssemblyName("MyAssembly"), typeof(object).Assembly);
ModuleBuilder mb = ab.DefineDynamicModule("MyModule");
TypeBuilder tb = mb.DefineType("MyType", TypeAttributes.Public);
TypeBuilder exceptionType = mb.DefineType("MyException", TypeAttributes.Public, typeof(Exception));
MethodBuilder method = tb.DefineMethod("Method", MethodAttributes.Public | MethodAttributes.Static, typeof(int), [typeof(int), typeof(int)]);
ILGenerator ilGenerator = method.GetILGenerator();
ilGenerator.BeginExceptionBlock();
ilGenerator.Emit(OpCodes.Ldarg_0);
ilGenerator.Emit(OpCodes.Ldarg_1);
ilGenerator.Emit(OpCodes.Add);
ilGenerator.BeginCatchBlock(exceptionType);  // Fails to begin catch block
ilGenerator.Emit(OpCodes.Ldc_I4_0);
ilGenerator.EndExceptionBlock();
ilGenerator.Emit(OpCodes.Ret);
tb.CreateType();
exceptionType.CreateType();
```

The current solution for handling unpopulated tokens: `BlobBuilder.ReserveBytes(int)` will not work for this case as `ControlFlowBuilder.AddCatchRegion(...)` validates and reserves the exception type handle, not writes the token into BLOB directly

### PR DESCRIPTION
**Pull request by buyaa-n**
PersistedAssemblyBuilder's `ILGenerator` cannot handle catch block with newly created `TypeBuilder` exception type, because the `TypeBuilder` token is not valid when `ILGenerator.BeginCatchBlock(Type? exceptionType)` called and the handle will not be populated properly until the assembly saved into a file or stream.

The current solution for handling unpopulated tokens: `BlobBuilder.ReserveBytes(int)` will not work for this case as `ControlFlowBuilder.AddCatchRegion(...)` validates and reserves the exception type handle, not writes the token into BLOB directly

So instead of adding exception handler blocks directly into the `ControlFlowBuilder` instance collecting them into a list and adding them to the `ControlFlowBuilder` during Save where all tokens will be populated. 

Alternative solution: tried to fill valid `TypeBuilder` handles using reflection, but it's getting very messy, need to access and manipulate private/internal members of `ControlFlowBuilder`.

Fixes https://github.com/dotnet/runtime/issues/106485

### CODE REVIEW

File: src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/PersistedAssemblyBuilder.cs [hunk 1]
```diff
@@ -202,5 +202,7 @@ protected override void SetCustomAttributeCore(ConstructorInfo con, ReadOnlySpan
         public override Module ManifestModule => _module ?? throw new InvalidOperationException(SR.InvalidOperation_AModuleRequired);
 
         public override AssemblyName GetName(bool copiedName) => (AssemblyName)_assemblyName.Clone();
+
+        public override bool IsDynamic => true;
```
Comment (jkotas) [inline src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/PersistedAssemblyBuilder.cs:206]: Why do we need this?

Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/PersistedAssemblyBuilder.cs:206]: It is unrelated with this PR, but saw it populated in the [missing docs](https://github.com/dotnet/runtime/issues/105982) list and thought is a public API that needs to be overriden

Comment (jkotas) [reply src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/PersistedAssemblyBuilder.cs:206]: I think we should delete from the ref instead. The abstract AssemblyBuilder provides this override already.

**Review by jkotas — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**